[#189] Implement custom table key sorting

ramanshah opened the pull request:
@chshersh As discussed, this code does not yet compile, but I'm opening this PR to facilitate discussion. Thanks!

jiegillet commented:
I agree with you assessment of the situation in the discussion. The previous ordering worked both for the lists of key/value pairs `[(Key, v)]` and for ordering the tables `[(Piece, v)]`, but if you want a specific ordering, you need to have two orderings and split `mapOrdered` in `mapOrderedKV` and `mapOrderedTables` or something like that.

chshersh commented on src/Toml/Printer.hs:
```diff
@@ -171,13 +173,17 @@ prettyTableArrays options i pref = mapOrdered arrText options
-mapOrdered f options
-    | shouldSort options = concatMap f . sortOn fst . HashMap.toList
-    | otherwise          = concatMap f . HashMap.toList
+mapOrdered :: ((Key, v) -> [t]) -> PrintOptions -> HashMap Key v -> [t]
```
Okay, I see the problem now. I don't think we should introduce a separate comparator for `Key` and `Piece`, because `Piece` can be considered as a special case of `Key`. So here in this function we can pass a list of type `[(Key, v)]` instead of `HashMap Key v` and convert to this list on all call-sites.

ramanshah replied:
Awesome - this was exactly the insight I was needing! Let me work on fleshing it out :)

chshersh commented on src/Toml/Printer.hs:
```diff
@@ -171,13 +173,17 @@ prettyTableArrays options i pref = mapOrdered arrText options
-    | otherwise          = concatMap f . HashMap.toList
+mapOrdered :: ((Key, v) -> [t]) -> PrintOptions -> HashMap Key v -> [t]
+mapOrdered f options = case printOptionsSorting options of
+    Just sorter -> concatMap f . sortBy (applyToFirst sorter) . HashMap.toList
```
You don't need to write `applyToFirst`, you can use standard function `comparing` here:

* https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Ord.html#v:comparing

ramanshah replied:
Perhaps I'm missing something? `comparing` doesn't accept a user-provided `a -> a -> Ordering` sorter. I couldn't find anything in the standard library that applies a custom sorter to tuples, either. (I attempted at some length to discover some slick `Functor` or `Applicative` thing to "lift" the user-provided sorter to tuples, but I failed and figured this one-line helper would be more readable anyway.)

chshersh replied:
@ramanshah ah, sorry, that was an oversight from my side. Here you indeed can't use `comparing`. I also don't see any elegant implementation using some standard functions. In that case, I can only recommend changing the type signature of the `applyToFirst` function further using _generalisation programming pattern_ in Haskell to the following (so there is only one possible implementation without any chance to introduce error):

```haskell
applyToFirst :: (a -> b -> c) -> (a, x) -> (b, y) -> c
```

ramanshah replied:
Very cool idea. I'm learning a lot on in this PR :)

chshersh commented on src/Toml/Printer.hs:
```diff
@@ -171,13 +173,17 @@ prettyTableArrays options i pref = mapOrdered arrText options
 -- Returns an indentation prefix
 tabWith :: PrintOptions -> Int -> Text
-tabWith options n = Text.replicate (n * indent options) " "
+tabWith options n = Text.replicate (n * printOptionsIndent options) " "
```
We are using `RecordWildCards` so such code usually can be written a little bit nicer 🙂 I think it looks a bit cleaner.
```suggestion
tabWith PrintOptions{..} n = Text.replicate (n * printOptionsIndent) " "
```

chshersh commented on src/Toml/Printer.hs:
```diff
@@ -26,17 +26,19 @@ import qualified Data.Text as Text
+    {- | How table keys should be sorted, if at all. -}
+    { printOptionsSorting :: !(Maybe (Key -> Key -> Ordering))
+    {- | Number of spaces by which to indent. -}
+    , printOptionsIndent  :: !Int
```
Oh, I just noticed. Documentation here is formatted a little bit wrong. It should be:

```haskell
{- | Configures the pretty printer. -}
data PrintOptions = PrintOptions
    { -- | How table keys should be sorted, if at all. 
    , printOptionsSorting :: !(Maybe (Key -> Key -> Ordering))

      -- | Number of spaces by which to indent.
    , printOptionsIndent  :: !Int
    }
```

ramanshah commented:
I got us a clean compilation and clean test suite!

I'm going to do a bit of refactoring (to respect the requested type signature for `mapOrdered`) and add test coverage of the custom sorting feature. Stay tuned. Are there any docs I need to build or update?

chshersh commented on src/Toml/Printer.hs:
```diff
@@ -171,13 +179,17 @@ prettyTableArrays options i pref = mapOrdered arrText options
-mapOrdered f options
-    | shouldSort options = concatMap f . sortOn fst . HashMap.toList
-    | otherwise          = concatMap f . HashMap.toList
+mapOrdered :: ((Key, v) -> [t]) -> PrintOptions -> HashMap Key v -> [t]
```
`mapOrdered` function should take `[(Key, v)]` instead of `HashMap`. This will make the function more performant. In the implementation of this function you're calling `toList` function anyways. But when you convert `HashMap Piece v` to `HashMap Key v` you spend extra time converting back to `HashMap` which is redundant, since you are going to work with the list anywas, so you can convert `HashMap Piece v` to `[(Key, v)]`

ramanshah replied:
Got it. Sorry, this code wasn't ready for re-review :) I will definitely do this!

chshersh commented on src/Toml/Printer.hs:
```diff
@@ -128,8 +130,14 @@ prettyKeyValue options i = mapOrdered (\kv -> [kvText kv]) options
+    asKeys = HashMap.fromList . map pieceToKey $ HashMap.toList asPieces
+
+    pieceToKey :: (Piece, a) -> (Key, a)
+    pieceToKey (piece, x) = (Key (pure piece), x)
```
You can implement `pieceToKey` using `first` function from `Data.Bifunctor`. There is `Bifunctor` instance for tuple.

ramanshah replied:
Learned something new here :)

ramanshah commented:
All set! I addressed the above comments and added a "golden" test of the custom sorting. I'm happy to add documentation as needed (and also would love comments on how to write a custom `Key -> Key -> Ordering` sorter nicely).

chshersh commented on test/Test/Toml/Printer/Golden.hs:
```diff
@@ -34,17 +35,31 @@ example = mkToml $ do
+    fromText x = Key $ pure $ Piece x
+
+spamEgg :: Key -> Key -> Ordering
+spamEgg k1 k2 = compare (spamEggDecorate k1) (spamEggDecorate k2)
```
This function can be written a little bit shorter:
```suggestion
spamEgg = comparing spamEggDecorate
```

chshersh commented on test/Test/Toml/Printer/Golden.hs:
```diff
@@ -34,17 +35,31 @@ example = mkToml $ do
+spamEggDecorate :: Key -> (Int, Key)
+spamEggDecorate k
+    | k == fromText "spam" = (0, fromText "spam")
+    | k == fromText "egg" = (1, fromText "egg")
```
You don't need to call `fromText` function here. There is `IsString` instance for `Key`. And we have `OverloadedStrings` language extension enabled project-wide. So you can just write the following and it will work:

```haskell
k == "egg" = (1, "egg")
```

ramanshah replied:
Oh that's so much better!

chshersh commented on src/Toml/Printer.hs:
```diff
@@ -128,8 +131,14 @@ prettyKeyValue options i = mapOrdered (\kv -> [kvText kv]) options
+    asKeys = map (first pieceToKey) $ HashMap.toList asPieces
+
+    pieceToKey :: Piece -> Key
+    pieceToKey piece = Key (pure piece)
```
I think eta-reduaction is okay here
```suggestion
    pieceToKey = Key . pure
```

ramanshah commented:
All set!

jiegillet commented on test/Test/Toml/Printer/Golden.hs:
```diff
@@ -34,17 +35,28 @@ example = mkToml $ do
+    | otherwise = (2, k)
+
+spamEgg :: Key -> Key -> Ordering
+spamEgg = comparing spamEggDecorate
```
Maybe a slightly more simple example of defining a custom ordering could be `spamEgg = compare on length` or something like that.

ramanshah replied:
Realistic use-cases of this feature involve key orders chosen semantically to serve human comprehension. They are usually arbitrary with respect to the key text. I'd like there to be some advertisement of how to to pretty-print a textually arbitrary key order. I think tuple decoration is a straightforward way to do so. This test is the only such advertisement so far. I'd personally have trouble reverse-engineering `compare \`on\` length` to design a semantically relevant sorter. I suffered an extreme amount of this type of obscurity among the Haskell YAML options, finally migrating a project to TOML because `tomland`'s codebase and docs are actually comprehensible. This type of comprehensibility matters a lot to me.

If it's important to keep the test code concise, we can go this way. But if so, let's discuss where to present a realistic pattern for a custom sort, either in the package documentation or in the blogosphere.

jiegillet replied:
I often use the pattern ``compare `on` f`` so to me the meaning is super clear: sort by string length. I understand if you are less used to it and you prefer more explicit, semantic functions, it's OK to do that. 

However, I feel a bit uneasy about the signature `spamEggDecorate :: Key -> (Int, Key)`. You are doing a function to sort `Key`, so you transform it to an `Int` but you still keep the `Key`, if two keys have the same integer value, you default to alphabetical order. I guess it's not such a big deal, but maybe `spamEggDecorate :: Key -> Int` would be more clear? It's up to you.

I can't really think of an ordering I'd want more than plain alphabetical to be honest, so I'll leave it up to you. As far as I'm concerned, the current PR is great.

ramanshah replied:
I grant that `(Int, Key)` has more cardinality than actually used in the sort.

The common use-case that motivated this feature is when one field in an array of tables/objects is critical for comprehension but gets buried in an alphabetical pretty-printing. Using an example from the `toml` readme, an alphabetically sorted array of tables might look like this:

```
[[products]]
name = "Hammer"
sku = 738594937

[[products]]
color = "gray"
name = "Nail"
sku = 284758393
```

Putting the name up front makes the table easier for a human to scan:

```
[[products]]
name = "Hammer"
sku = 738594937

[[products]]
name = "Nail"
color = "gray"
sku = 284758393
```

Realistic data structures tend to suffer from a more extreme version of this problem because they tend to be bigger, often with substantial nesting; I've found alphabetically pretty-printed data structures over the course of my career surprisingly hard to read. I frequently want "alphabetical with exceptions" to make pretty-printed data more legible, and the given sorter implements this.

For the purpose of the test suite, the given sorter also corresponds to a minimal change in the golden file (transposing two lines of TOML compared to the alphabetical sort, which I was able to construct by hand), helping to verify correctness of the library code.

jiegillet replied:
I see, you make a good case. For the purpose of wanting to get `name` on top, `(Int, Key)` is actually nice since anything non-covered explicitly will be alphabetical sorted. 

Thank you for sharing your insights on realistic uses of config files, that is helpful to me.

ramanshah commented on src/Toml/Printer.hs:
```diff
@@ -171,13 +180,17 @@ prettyTableArrays options i pref = mapOrdered arrText options
-    | otherwise          = concatMap f . HashMap.toList
+mapOrdered :: ((Key, v) -> [t]) -> PrintOptions -> [(Key, v)] -> [t]
+mapOrdered f options = case printOptionsSorting options of
+    Just sorter -> concatMap f . sortBy (applyToFirst sorter)
```
Oh very nice! That's the elegant thing we were looking for previously :)

jiegillet replied:
Right, I missed that discussion. Well, better late than never ^^
